[Merged by Bors] - feat: local frames in a vector bundle

grunweg opened the pull request:
We construct local frames on a vector from local trivialisations. The former constructs, given a trivialisation `e` of the vector bundle and a basis of the model fibre, `Basis.localFrame` is a local frame on `e.baseSet`
A future PR will use this to define the local extension of a tangent vector to a vector field near a point.

From the path towards geodesics and the Levi-Civita connection.

Co-authored-by: Patrick Massot [patrickmassot@free.fr](mailto:patrickmassot@free.fr)

---

- [x] depends on: #27021
- [x] depends on: #30338
- [x] depends on: #31542
- [x] depends on: #31788

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg committed:
feat: local frames

This depends on the custom elaborators added in #27021;
for some reason, using these fails. To be investigated!

grunweg committed:
Update for recent changes (to precedence of T%)

grunweg committed:
Fix the build

grunweg committed:
Polish

grunweg committed:
Fix the build

grunweg committed:
Extend doc-string; remove local extension stuff for now

grunweg committed:
Remove another lemma

grunweg committed:
Merge master

grunweg committed:
repr -> coeff

grunweg committed:
Last rename, line length and variable management

grunweg committed:
Comment

grunweg committed:
Tweak

ocfnash commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -283,3 +301,334 @@ end
 
 end IsLocalFrame
+
+namespace Module.Basis
```
I'm not absolutely convinced about all these results going into the `Module.Basis` namespace. What do you think about putting them in `Trivialization`.

(Aside: IMHO `Trivialization` itself should be in the `Bundle` namespace but this is a question for another day.)

ocfnash commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -283,3 +301,334 @@ end
+lemma localFrame_toBasis_at_coe
+    (e : Trivialization F (Bundle.TotalSpace.proj : Bundle.TotalSpace F V → M))
+    [MemTrivializationAtlas e] (b : Basis ι 𝕜 F) (i : ι) (hx : x ∈ e.baseSet) :
+    b.localFrame_toBasis_at e hx i = b.localFrame e i x := by simp [hx]
```
Do we definitely want this lemma?

It's just `(b.localFrame_apply_of_mem_baseSet e hx).symm`.

grunweg replied:
Fair point: let me remove it for now.

ocfnash commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -283,3 +301,334 @@ end
+
+omit [IsManifold I 0 M] in
+variable (e b) in
+@[simp]
```
Again, I'm dubious about these conditional `simp` lemmas. (I'll stop raising this point; perhaps they're a good idea.)

ocfnash commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -283,3 +301,334 @@ end
+    b.localFrame_coeff I e i s x = (b.localFrame_toBasis_at e hx).repr (s x) i := by
+  have ilf := b.localFrame_isLocalFrameOn_baseSet I 1 e
+  rw [show localFrame_toBasis_at e b hx = ilf.toBasisAt hx by ext j; simp [localFrame, hx]]
+  exact ilf.coeff_apply_of_mem hx s i -- XXX better variable name!
```
What does this comment mean?

ocfnash commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -283,3 +301,334 @@ end
+  let basL : F →L[𝕜] 𝕜 := {
+    toLinearMap := basl
+    cont := basl.continuous_of_finiteDimensional
+  }
```
Minor style nits:
```suggestion
  let basl : F →ₗ[𝕜] 𝕜 :=
    { toFun := bas
      map_add' m m' := by simp [bas]
      map_smul' m x := by simp [bas] }
  let basL : F →L[𝕜] 𝕜 :=
    { toLinearMap := basl
      cont := basl.continuous_of_finiteDimensional }
```

ocfnash reviewed (commented):
Thanks!

I've left a bunch of fairly superficial comments, but I think they're worth addressing. I'll return again once you've had a chance to respond.

grunweg commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -283,3 +301,334 @@ end
+  if hx : x ∈ e.baseSet then b.localFrame_toBasis_at e hx i else 0
+
+-- TODO: understand why this isn’t already a simp lemma
+attribute [simp] Trivialization.apply_mk_symm
```
Let me try that: #31127

grunweg committed:
Style tweaks from review

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

grunweg committed:
Apply suggestions from code review

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

grunweg committed:
Polish

grunweg committed:
Merge branch 'master' into localframes

grunweg committed:
Adapt

grunweg committed:
chore: clean up variables

grunweg committed:
Rename

grunweg committed:
Move to Trivialization namespace

grunweg committed:
Unsimp one lemma

grunweg commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -283,3 +301,334 @@ end
+    CMDiffAt n (T% (b.localFrame e i)) x :=
+  (b.localFrame_isLocalFrameOn_baseSet I n e).contMDiffAt e.open_baseSet hx _
+
+@[simp]
```
I've thought about this for a fair bit now: there other condition simp lemmas about trivialisations, so this is not without precedent. I think that one would only want to rewrite with these lemmas when a hypothesis `x \in e.baseSet` is in scope anyway, so I think this should be fine.
That said, unsimping all these lemmas is also fairly painless (on Patrick's and my branch) - so if you feel strongly, I can remove the simp tag also.

I've un-simped the lemma with the other hypothesis: I don't see it being used much.

grunweg committed:
More review comments: remove almost-duplicate lemma; fix module docs for rename, and add two missing doc-strings

grunweg committed:
Cleaner line breaks

ocfnash commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -84,23 +104,21 @@ open scoped Bundle Manifold ContDiff
   [∀ x : M, TopologicalSpace (V x)]
-  [FiberBundle F V] [VectorBundle 𝕜 F V] [ContMDiffVectorBundle n F V I]
-  -- `V` vector bundle
+  -- `V` a fibre bundle/vector bundle over `F` (the latter is needed only in some later sections)
```
I find this slightly confusing; let's just drop it.
```suggestion
```

ocfnash commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -283,3 +301,302 @@ end
+  by_cases hxe : x ∈ e.baseSet
+  · simp [localFrame_coeff, hxe]
+    congr
+  · simp [localFrame_coeff, hxe]
```
```suggestion
  by_cases hxe : x ∈ e.baseSet <;> aesop
```

ocfnash commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -283,3 +301,302 @@ end
+variable (e b) in
+lemma localFrame_coeff_apply_zero_at (hs : s x = 0) (i : ι) : e.localFrame_coeff I b i s x = 0 := by
+  simp only [localFrame_coeff]
+  exact (localFrame_isLocalFrameOn_baseSet I 1 e b).coeff_apply_zero_at hs i
```
I'm dubious about this lemma. It's a trivial consequence of linearity so I feel something is wrong with the API if we need it.

ocfnash commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -283,3 +301,302 @@ end
+  have hbas : ContMDiffAt 𝓘(𝕜, F) 𝓘(𝕜) k basL (e (s x)).2 :=
+    contMDiffAt_iff_contDiffAt.mpr <| (basL.contDiff (n := k)).contDiffAt
+  exact hbas.comp x h₁
+
```
```suggestion

variable (b)

```
and then drop `(b : Basis ι 𝕜 F)` from numerous lemmas below

grunweg committed:
Drop Fintype \iota assumptions, and ask for finite-dimensionality of F instead.

Over vector bundles, these are equivalent (and this allows removing
essentially duplicate hypotheses from theorems).

grunweg committed:
Apply suggestions from code review

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

grunweg committed:
More variable clean-up

grunweg committed:
Use more variable

grunweg committed:
Use more variable

grunweg committed:
Needs more variable

grunweg committed:
chore: golf one proof

grunweg committed:
Rename

grunweg committed:
Update Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

grunweg committed:
One more name

grunweg committed:
Add contMDiffOn_empty and friends

grunweg committed:
Also tag with fun_prop

grunweg committed:
Golf using new lemmas

grunweg committed:
Merge branch 'master' into localframes

ocfnash commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -235,26 +269,34 @@ lemma contMDiffOn_of_coeff [Fintype ι] (h : ∀ i, CMDiff[u] n (hs.coeff i t))
+  · -- future: could push replace the additional lemma?
+    -- TODO: why does `push_neg at hU` make `hU disappear from the local context?
+    simp [Set.not_nonempty_iff_eq_empty'.mp hu]
+  obtain ⟨x, hx⟩ := hu
```
```suggestion
  rcases u.eq_empty_or_nonempty with rfl | ⟨x, hx⟩; · simp
```

ocfnash commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -283,3 +326,258 @@ end
+
+variable [ContMDiffVectorBundle 1 F V I]
+
+open scoped Classical in
```
This seems redundant:
```suggestion
```

ocfnash reviewed (commented):
How about we split out a PR with all the content strictly before `localFrame_coeff`: I think this is all ready to go (modulo the minor nitpicks I've raised as inline comments).

With that done, I think it probably is worth continuing by exploring ideas discussed [here](https://github.com/leanprover-community/mathlib4/pull/30083/files#r2515218603).

grunweg commented:
Sounds good! I'll fix the easy comments you mentioned, fix the typo in the module doc-string and make a separate PR with just the first part.

grunweg committed:
Rename and fix docs typo

grunweg committed:
Apply suggestions from code review

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

grunweg commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -283,3 +326,258 @@ end
+  simp [localFrame, hx]
+
+lemma localFrame_apply_of_notMem {i : ι} (hx : x ∉ e.baseSet) : e.localFrame b i x = 0 := by
+  simp [localFrame, hx]
```
Indeed, that's nicer. Thanks!

grunweg committed:
Reorder for nicer proofs

grunweg committed:
Merge branch 'master' into localframes

grunweg committed:
Add foo; TODO find a good name!

grunweg committed:
WIP: untested adaptations

grunweg committed:
Merge branch 'master' into localframes

michaellee94 committed:
Fix the build

Co-authored by:  michaellee94 <michael.a.rodrigues.lee@gmail.com>

grunweg committed:
Clean-up

grunweg committed:
wip: redefine coeff per review

grunweg committed:
wip: redefine coefficients

michaellee94 committed:
Finish migration of `LocalFrame.lean`, address some TODOs (#5)

grunweg committed:
Tweaks

grunweg commented:
@ocfnash This PR should be ready for review again - all review comments are addressed now.
Thanks a lot, @michaellee94, for your help with that!

grunweg committed:
Simplify and fix

grunweg committed:
A few more; not exhaustive

grunweg committed:
A few more

grunweg committed:
Last ones

ocfnash commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -178,101 +187,103 @@ alias fintype_of_finiteDimensional := fintypeOfFiniteDimensional
-  have := fintypeOfFiniteDimensional hs hx
-  exact ⟨fun h i ↦ hs.coeff_congr h i, fun h ↦ by
-    simp +contextual [h, hs.coeff_sum_eq t hx, hs.coeff_sum_eq t' hx]⟩
+    t x = t' x ↔ ∀ i, (hs.coeff i x) (t x) = (hs.coeff i x) (t' x) := by
```
```suggestion
    t x = t' x ↔ ∀ i, hs.coeff i x (t x) = hs.coeff i x (t' x) := by
```

ocfnash commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -178,101 +187,103 @@ alias fintype_of_finiteDimensional := fintypeOfFiniteDimensional
 lemma coeff_apply_zero_at (hs : IsLocalFrameOn I F n s u) (ht : t x = 0) (i : ι) :
-    hs.coeff i t x = 0 := by
-  simp [hs.coeff_congr (t' := 0) ht]
+    (hs.coeff i x) (t x) = 0 := by simp [ht]
```
I'll stop with these very nitpicking "remove brackets" suggestions now but it would be nice to try to kill all of these.
```suggestion
    hs.coeff i x (t x) = 0 := by simp [ht]
```

grunweg replied:
I think I caught them all - good catch!

ocfnash commented on Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean:
```diff
@@ -369,6 +383,206 @@ lemma _root_.contMDiffAt_localFrame_of_mem (i : ι) (hx : x ∈ e.baseSet) :
+variable [VectorBundle 𝕜 F V] [ContMDiffVectorBundle 1 F V I]
+  {e : Trivialization F (TotalSpace.proj : TotalSpace F V → M)} [MemTrivializationAtlas e]
+  {ι : Type*} (b : Basis ι 𝕜 F) {s : Π x : M, V x} {t : Set M} {k : WithTop ℕ∞} {x x' : M}
+variable [FiniteDimensional 𝕜 F] [CompleteSpace 𝕜] [ContMDiffVectorBundle k F V I]
```
```suggestion
  [FiniteDimensional 𝕜 F] [CompleteSpace 𝕜] [ContMDiffVectorBundle k F V I]
```

ocfnash reviewed (approved):
Thanks!

bors d+

grunweg committed:
Some review comments

grunweg committed:
More

grunweg committed:
Update Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

grunweg committed:
Update Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

grunweg committed:
Tweak

grunweg committed:
Last parentheses
